### PR TITLE
New flexible MLP; a small bug fix; some clean up 

### PR DESCRIPTION
**Pull request by nkrah**
Mainly enhancement implementing new flexible MLP which works for any proton energy and does not rely on external parametrization. Some clean up along the way.

### CODE REVIEW

File: pctPolynomialMLPFunction.txx [hunk 1]
```diff
@@ -23,10 +24,45 @@ PolynomialMLPFunction
...
+    case 4:
+      m_bm.reserve(Functor::PolynomialMLP::bm_4.size());
+      std::copy(Functor::PolynomialMLP::bm_4.begin(),Functor::PolynomialMLP::bm_4.end(),std::back_inserter(m_bm));
+      break;
+    case 5:
+      m_bm.reserve(Functor::PolynomialMLP::bm_5.size());
+      std::copy(Functor::PolynomialMLP::bm_5.begin(),Functor::PolynomialMLP::bm_5.end(),std::back_inserter(m_bm));
+      break;
+    default:
+      std::cerr << "Allowed values for polydeg are 0-5. Received " << m_PolynomialDegree << ". Using default (5)." << std::endl;
```
Comment (SimonRit) [inline pctPolynomialMLPFunction.txx:55]: Maybe use an `itkWarningMacro`

File: pctSchulteMLPFunction.txx [hunk 2]
```diff
@@ -69,6 +69,13 @@ SchulteMLPFunction
 
 }
 
+void
+SchulteMLPFunction
+::Init(const VectorType posIn, const VectorType posOut, const VectorType dirIn, const VectorType dirOut, double eIn, double eOut)
+{
+  itkGenericExceptionMacro("This version of the Init method not implemented for derived class SchulteMLPFunction.");
```
Comment (SimonRit) [inline pctSchulteMLPFunction.txx:76]: Why not defining it in the mother class as default?

File: pctThirdOrderPolynomialMLPFunction.txx [hunk 3]
```diff
@@ -6,9 +6,16 @@ void
 ThirdOrderPolynomialMLPFunction<TCoordRep>
 ::InitUncertain(const VectorType posIn, const VectorType posOut, const VectorType dirIn, const VectorType dirOut, double dEntry, double dExit, double m_TrackerResolution, double m_TrackerPairSpacing, double m_MaterialBudget)
 {
-  std::cout << "Not implemented for this derived class." << std::endl;
+  itkGenericExceptionMacro("InitUncertain not implemented for this derived class ThirdOrderPolynomialMLPFunction.");
```
Comment (SimonRit) [inline pctThirdOrderPolynomialMLPFunction.txx:9]: Why not defining it in the mother class as default?

**Review by SimonRit — CHANGES_REQUESTED** [review]
Looks good in general. I would suggest to rename it, I find flexible a bit vague. Why not EnergyLoss? Or EnergyAdaptive?

File: pctPolynomialMLPFunction.txx (hunk 1 above)
Comment (nkrah) [reply pctPolynomialMLPFunction.txx:55]: Yes, better.

**Comment by nkrah** [comment]
Implemented the suggested changes.

File: pctPolynomialMLPFunction.txx [hunk 4]
```diff
@@ -233,52 +118,6 @@ PolynomialMLPFunction
...
-  }
-
-  x += m_dm_x[0];
-  y += m_dm_y[0];
-
-
-#ifdef MLP_TIMING
-  m_EvaluateProbe1.Stop();
-#endif
 }
```
Comment (SimonRit) [inline pctPolynomialMLPFunction.txx:121]: Why was this deleted? Intentional?

**Review by SimonRit — APPROVED** [review]

File: pctPolynomialMLPFunction.txx (hunk 4 above)
Comment (nkrah) [reply pctPolynomialMLPFunction.txx:121]: That was a version of the function which I had implemented in the beginning. It calculates the MLP at a single depth and not a whole array of depth. The vectored version of the function is much faster. The DistanceDrivenProjection only uses the vectored version so I saw no motivation to keep the old one.
